### PR TITLE
feat: cross-platform light/dark mode API

### PR DESCRIPTION
**Pull request by sgtcoolguy**
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-27501

**Description:**
This incorporates the code from @drauggres in his PR #11301

This attempts to make a more comprehensive cross-platform API for dealing with determining the current theme/mode, and consistency in the return value of `Ti.UI.fetchSemanticColor`.

**API**
- `Ti.UI.userInterfaceStyle` reports the current theme via `Ti.UI` constants: `USER_INTERFACE_STYLE_DARK`, `USER_INTERFACE_STYLE_LIGHT`, and `USER_INTERFACE_STYLE_UNSPECIFIED`
- `Ti.UI.fetchSemanticColor(String name)` returns a `Ti.UI.Color` proxy for the named color (and you can get the underlying "color" by calling `toHex()` on it to get a hex string) on iOS 13+, a string (representing the hex value) on Android and iOS < 13. The mismatched return types are necessary to make it easy to just pass the returned value to the UI properties without conversion right now.
  - Note that we support usage of the "standard" named colors baked into iOS that are adaptive to theme changes: [Standard Colors](https://developer.apple.com/documentation/uikit/uicolor/standard_colors?language=objc) and [UI Element Colors](https://developer.apple.com/documentation/uikit/uicolor/ui_element_colors?language=objc)
- `Ti.UI` now has a new `userInterfaceStyle` event which gets fired when the device theme changes. The object received in the event listener has a `value` property whose value is the constant for the new `USER_INTERFACE_STYLE_*` that was set.

**Deprecations**
- `Ti.App.iOS.userInterfaceStyle` -> use `Ti.UI.userInterfaceStyle`
- `Ti.App.iOS.USER_INTERFACE_STYLE_DARK` -> use `Ti.UI.USER_INTERFACE_STYLE_DARK`
- `Ti.App.iOS.USER_INTERFACE_STYLE_LIGHT` -> use `Ti.UI.USER_INTERFACE_STYLE_LIGHT`
- `Ti.App.iOS.USER_INTERFACE_STYLE_UNSPECIFIED` -> use `Ti.UI.USER_INTERFACE_STYLE_UNSPECIFIED`
- Undocumented `Ti.UI.iOS.fetchSemanticColor()` (use `Ti.UI.fetchSemanticColor()`)

**Doc updates**
- Document the `Ti.UI.Color` proxy. This is really an iOS-only proxy. We need a way to pass around the underlying `UIColor` objects from iOS so that we can set various UI color properties with them and they get "automatically" updated if the user changes the device theme.

**Notable parity issues**
While passing around `Ti.UI.Color` proxies on iOS allows for the UI to automatically adjust to theme changes, we don't have the equivalent on Android. Can we hook an equivalent `Ti.UI.Color` proxy that could auto-update on Android?

**To Test**
The actual API has some unit tests to verify the new properties/constants exist.

It's difficult to test on the emulator/sim, but to test the new event on theme changes you can run this app:
### Resources/semantic.colors.json ###
```json
{
	"primaryBackground": {
		"light": "#efece3",
		"dark": "#001900"
	}
}
```
### Resources/app.js ###
```javascript
// Also try baked-in colors on iOS like 'systemRedColor' - they will change as light/dark mode changes - although very subtly (I had to use the Digital Color Meter to double-check)
const primaryBackground = Ti.UI.fetchSemanticColor('primaryBackground');
const win = Ti.UI.createWindow({
    backgroundColor: primaryBackground
});

const btn = Ti.UI.createButton({
    title: 'Trigger'
});

btn.addEventListener('click', function() {
    Ti.API.info('Hello world!');
});

win.add(btn);
win.open();

Ti.UI.addEventListener('userinterfacestyle', e => {
    console.log(e);
    console.log(primaryBackground.toHex());
});
```
On iOS Emulator, I had to use the **Hardware > Siri** menu and then tell Siri to "Tun Dark mode on" (or off). On Android, You need an apilevel 29+ emulator, and you can drag from he top status bar down to get a quick menu. Click the pencil icon to edit the options displayed, drag the "Dark Theme" button up into it, and then you can use that to toggle dark/light themes while the app is running.

### CODE REVIEW

File: android/cli/commands/_build.js [hunk 1]
```diff
@@ -3406,6 +3409,101 @@ AndroidBuilder.prototype.generateI18N = async function generateI18N() {
 	}
 };
 
+AndroidBuilder.prototype.generateSemanticColors = async function generateSemanticColors() {
+	this.logger.info(__('Generating semantic colors resources'));
+	const _t = this;
+	const xmlFileName = 'semantic.colors.xml';
+	const valuesDirPath = path.join(this.buildAppMainResDir, 'values');
+	const valuesNightDirPath = path.join(this.buildAppMainResDir, 'values-night');
```
Comment (jquick-axway) [inline android/cli/commands/_build.js]: Don't create a "values.xml" file. We risk file collision with the app developer's own "res" files.
They need to be prefixed with "ti_" or "ti-". (C-style convention.)

So, we should name the files "ti_semantic_values.xml" and "ti_semantic_values_night.xml".
Or "ti-semantic-values.xml" and "ti-semantic-values-night.xml".

Note: Drawable "res" files can't have dash '-' chars (they're illegal), but XML "res" files can.

**Comment by jquick-axway** [comment]
> Do we add an event to Ti.UI to fire when theme changes and let developers decide if they want to try and handle it?

On Android, we know when the system changes between dark/light theme on Android Q (and day/night mode on older OS versions) when Java method `onConfigurationChanged()` gets invoked with "uiMode" set.

> Can we hook an equivalent Ti.UI.Color proxy that could auto-update on Android?

Theoretically yes. Currently, colors are applied to view via the Java [ColorDrawable](https://developer.android.com/reference/android/graphics/drawable/ColorDrawable) class which is assigned a hard-coded color value. We could create our own `TiColorDrawable` which can dynamically change colors internally. This would involve us setting up a global listener for when "uiMode" changes, and when it does, our custom `TiColorDrawable` would have to call its `invalidateSelf()` method, which signals the view to re-render itself with the updated drawable.

---
The only thing that bugs me here is that we're conflating dark/light theme with day/night mode.

**Comment by sgtcoolguy** [comment]
@jquick-axway Is there really a difference between Day/Night and Light/Dark modes? It seems to me to be a re-branding from Android: https://developer.android.com/guide/topics/ui/look-and-feel/darktheme
I mean in effect, they use both Light/Dark theme and Day/Night more or less interchangeably. The docs there even show that `NIGHT_MODE_NO` == Light and `NIGHT_MODE_YES` == Dark.

**Comment by jquick-axway** [comment]
Right, I've seen Google's docs on this. I don't know how they're differentiating the 2 concepts yet. I haven't had time to play with it.

On Android, if you drag the top status bar down and tap on the "Night Light" button (ie: the "Moon" button), that will flip the Day/Night mode. This triggers the "uiMode" config-change. This is not intended to trigger Dark mode (but I suppose you can set up an app to do this). Changing Dark/Light theme under Display settings is supposed to trigger "uiMode" as well. So, we need to figure out how to differentiate between Day/Night and Dark/Light changes.

**Comment by jquick-axway** [comment]
I think you have to use the [MODE_NIGHT_FOLLOW_SYSTEM](https://developer.android.com/reference/androidx/appcompat/app/AppCompatDelegate#MODE_NIGHT_FOLLOW_SYSTEM) constant on Android Q (and above) to have the app honor the system's light/dark theme (not day/night setting). We would also have to update Titanium's app theme to "Theme.AppCompat.DayNight".

File: apidoc/Titanium/UI/UI.yml [hunk 2]
```diff
@@ -174,6 +174,16 @@ description:  |
 extends: Titanium.Module
 since: "0.4"
 
+events:
+  - name: userInterfaceStyle
+    summary: Fired when the `userInterfaceStyle` changes.
```
Comment (drauggres) [inline apidoc/Titanium/UI/UI.yml]: Shouldn't it be lowercase (`userinterfacestyle`)?

Comment (sgtcoolguy) [reply apidoc/Titanium/UI/UI.yml]: I used the camelCase name in the code. but it may be more consistent to use all lowercase?

Comment (drauggres) [reply apidoc/Titanium/UI/UI.yml]: https://jira.appcelerator.org/browse/TIMOB-10291

Comment (sgtcoolguy) [reply apidoc/Titanium/UI/UI.yml]: Man, you are really digging up some old tickets/references! 👍 ok, changing to all lowercase

File: android/cli/commands/_build.js (hunk 1 above)
Comment (sgtcoolguy) [reply android/cli/commands/_build.js]: @jquick-axway These aren't creating a values.xml file, they're creating:
- `build/app/src/main/res/values/semantic.colors.xml`
- `build/app/src/main/res/values-night/semantic.colors.xml`

Is the suggestion we should instead name them:
- `build/app/src/main/res/ti_semantic_colors.xml`
- `build/app/src/main/res/ti_semantic_colors_night.xml`
?

**Review by janvennemann — APPROVED** [review]
Just a minor note on the docs. The iOS changes are looking good, nothing to add there.

**Comment by jquick-axway** [comment]
I tested this PR with the below code on Android. This will automatically change the background color when flipping between Light/Dark mode. It also tests the handling of multiple windows.

```javascript
function createThemedWindow(settings) {
	if (!settings) {
		settings = {};
	}
	settings.backgroundColor = Ti.UI.fetchSemanticColor("primaryBackground");
	var window = Ti.UI.createWindow(settings);
	function onUserInterfaceStyleChanged(e) {
		Ti.API.info("@@@ userinterfacestyle event.value: " + e.value);
		window.backgroundColor = Ti.UI.fetchSemanticColor("primaryBackground");
	}
	window.addEventListener("open", function() {
		Ti.UI.addEventListener("userinterfacestyle", onUserInterfaceStyleChanged);
	});
	window.addEventListener("close", function() {
		Ti.UI.removeEventListener("userinterfacestyle", onUserInterfaceStyleChanged);
	});
	return window;
}

var parentWindow = createThemedWindow({ title: "Parent Window" });
var openButton = Ti.UI.createButton({ title: "Show Child" });
openButton.addEventListener("click", function() {
	var childWindow = createThemedWindow({ title: "Child Window" });
	var closeButton = Ti.UI.createButton({ title: "Close" });
	closeButton.addEventListener("click", function() {
		childWindow.close();
	});
	childWindow.add(closeButton);
	childWindow.open();
});
parentWindow.add(openButton);
parentWindow.open();
```

File: common/Resources/ti.internal/extensions/ti/ti.ui.js [hunk 3]
```diff
@@ -1,60 +1,71 @@
...
+				// We should probably throw an Error here (or return a fallback color!)
 				console.error('Failed to load colors file \'semantic.colors.json\'');
 				return;
 			}
 		}
+
 		try {
-			return colorset[colorName][uiModule.semanticColorType].color || colorset[colorName][uiModule.semanticColorType];
+			const entry = colorset[colorName][UI.semanticColorType];
+			const hex = entry.color || entry;
```
[the remark this answers is not in the document]
Comment (sgtcoolguy) [reply common/Resources/ti.internal/extensions/ti/ti.ui.js]: I believe @drauggres has an open PR for this at #11315

[the remark this answers is not in the document]
Comment (ewanharris) [reply common/Resources/ti.internal/extensions/ti/ti.ui.js]: Yes (halfbaked as most of it was), I believe that we handle the alpha property when generating the iOS semantic colors file, but not in the lookup in the JS code

**Comment by sgtcoolguy** [comment]
ok, I think I addressed the remaining review feedback here? I'm gonna tag for QE to test now...

File: common/Resources/ti.internal/extensions/ti/ti.ui.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jquick-axway) [reply common/Resources/ti.internal/extensions/ti/ti.ui.js]: Okay. I scheduled [TIMOB-27519](https://jira.appcelerator.org/browse/TIMOB-27519) for 9.1.0. Thanks

**Review by jquick-axway — APPROVED** [review]
CR: Pass

**Comment by ssekhri** [comment]
@sgtcoolguy, the event _userInterfaceStyle_ does not fire on iOS when the theme is changed. It does work on Android.
Tested on iOS device and simulator version 13.4.
Env:
Mac OS: 10.15.4
SDK: 9.1.0.v20200430122302
Appc CLI: 8.0.0
JDK: 11.0.4
Node: 10.17.0
Xcode: 11.4

**Comment by sgtcoolguy** [comment]
@ssekhri I forgot to update the code sample in the PR here. The event name changed from `userInterfaceStyle` to `userinterfacestyle` (all lowercase) - I forgot we had a naming scheme of using all lowercase for events (and was trying to match the property name).

FYI, for whatever reason this code did not work at all for me on Xcode 11.5 beta 1, but works on Xcode 11.4. It would return black for named colors for me. Here's hoping Xcode 11.5 beta 2 fixes it?

**Comment by ssekhri** [comment]
The fetchSemanticColor() method and userinterfacestyle event work fine on android and iOS for dark and light mode.
Verified on:
Mac OS: 10.15.4
SDK: 9.1.0.v20200430122302
Appc CLI: 8.0.0
JDK: 11.0.4
Node: 10.17.0
Xcode: 11.4
Device: iPhoneX(v13.4), iOS simulator 13.4, Pixel 3(v10.0), Pixel 2 emulator(v10.0)

**Comment by hansemannn** [comment]
A tiny addition to this great step forward: [TIMOB-27895](https://jira.appcelerator.org/browse/TIMOB-27895)

**Comment by KamalKant-Tech** [comment]
This is event is not triggering for me when I change the theme mode in android and iOS.

```
Ti.UI.addEventListener('userinterfacestyle', e => {
    console.log(e);
    console.log(primaryBackground.toHex());
});
```

**Comment by ewanharris** [comment]
@KamalKant-Tech, what SDK are you using? This will ship in SDK 9.1.0, and the event is firing ok for me when I use an SDK from that branch (`appc ti sdk install -b 9_1_X`). Before commenting on PR's, please make sure to check the JIRA ticket (usually linked in the description) and check the `Fix Version/s` field there against your SDK version

**Comment by KamalKant-Tech** [comment]
@ewanharris Thanks, I got the issue it was not working because of SDK version. I have been using 9.0.3 and then I moved to 9.1.0 now it works.

**Pull request merged**